ksql-engine/src/main/java/io/confluent/ksql/util/PersistentQueryMetadata.java

```diff
@@ -62,15 +61,28 @@ public PersistentQueryMetadata(final String statementString,
         closeCallback);
     this.id = Objects.requireNonNull(id, "id");
     this.resultTopic = Objects.requireNonNull(resultTopic, "resultTopic");
-    this.sinkNames = new HashSet<>();
-    this.sinkNames.add(sinkDataSource.getName());
+    this.sinkNames = ImmutableSet.of(sinkDataSource.getName());
 
     if (resultTopic.getKsqlTopicSerDe() == null) {
       throw new KsqlException(String.format("Invalid result topic: %s. Serde cannot be null.",
           resultTopic.getName()));
     }
   }
 
+  private PersistentQueryMetadata(
+      final PersistentQueryMetadata other,
+      final Consumer<QueryMetadata> closeCallback
+  ) {
```

Comment:
The sandbox needs to be able to terminate and create queries, hence it needs a way to clone the engine's queries when its created, installing a different `closeCallback` that updates the sandboxes state, not the main engine.  Hence this method...